### ISSUE TITLE
pg gem fails with Chef::Exceptions::ChildConvergeError: Chef run process terminated by signal 6 (IOT)

### ISSUE THREAD

**Issue by sekrett** [chef/chef#5582]
## Description

I am using 'database' cookbook with PostgreSQL v9.5.4. It works fine but later (I don't know what changes) it starts to fail with any operation with PostgreSQL. No problem with MySQL.

## Chef Version

12.16.42

## Platform Version

FreeBSD 10.3-RELEASE-p7 amd64

## Replication Case

I am running chef-client as daemon:
```
/opt/chef/embedded/bin/ruby /usr/bin/chef-client --daemonize --runlist recipe[bosting-generator] --pid /var/run/chef_client.pid
```

/etc/chef/client.rb:
```
node_name       'bosting-generator'
log_location    '/var/log/chef-client'
local_mode      true
cookbook_path   ['/root/cookbooks']
log_level       :debug
```

The simplest recipe ever what fails also:
```
%w(redis hiredis mysql2 pg).each do |name|
  chef_gem name do
    compile_time true
  end
  require name
end
```

I tried to delete /opt/chef, installed chef-client again, tried manually to install gems:
```
/opt/chef/embedded/bin/gem install mysql2 pg redis hiredis
```

It installs all gems fine but chef-client stills fails with the same error. Of course I restarted chef-client daemon, does not help. I rebooted the server, still does not help.

I tried:
```
/opt/chef/embedded/bin/gem pristine --all
```
Also does not help.

## Client Output

```
[2016-11-25T17:01:00+03:00] INFO: chef_gem[pg] installed pg at 0.19.0
[2016-11-25T17:01:00+03:00] DEBUG: chef_gem[pg] resetting gem environment to default
[2016-11-25T17:01:01+03:00] ERROR: Chef::Exceptions::ChildConvergeError: Chef run process terminated by signal 6 (IOT)
[2016-11-25T17:01:01+03:00] DEBUG: Chef::Exceptions::ChildConvergeError: Chef run process terminated by signal 6 (IOT)
/opt/chef/embedded/lib/ruby/gems/2.3.0/gems/chef-12.16.42/lib/chef/application.rb:317:in `handle_child_exit'
/opt/chef/embedded/lib/ruby/gems/2.3.0/gems/chef-12.16.42/lib/chef/application.rb:304:in `fork_chef_client'
/opt/chef/embedded/lib/ruby/gems/2.3.0/gems/chef-12.16.42/lib/chef/application.rb:247:in `block in run_chef_client'
/opt/chef/embedded/lib/ruby/gems/2.3.0/gems/chef-12.16.42/lib/chef/local_mode.rb:44:in `with_server_connectivity'
/opt/chef/embedded/lib/ruby/gems/2.3.0/gems/chef-12.16.42/lib/chef/application.rb:235:in `run_chef_client'
/opt/chef/embedded/lib/ruby/gems/2.3.0/gems/chef-12.16.42/lib/chef/application/client.rb:464:in `sleep_then_run_chef_client'
/opt/chef/embedded/lib/ruby/gems/2.3.0/gems/chef-12.16.42/lib/chef/application/client.rb:451:in `block in interval_run_chef_client'
/opt/chef/embedded/lib/ruby/gems/2.3.0/gems/chef-12.16.42/lib/chef/application/client.rb:450:in `loop'
/opt/chef/embedded/lib/ruby/gems/2.3.0/gems/chef-12.16.42/lib/chef/application/client.rb:450:in `interval_run_chef_client'
/opt/chef/embedded/lib/ruby/gems/2.3.0/gems/chef-12.16.42/lib/chef/application/client.rb:434:in `run_application'
/opt/chef/embedded/lib/ruby/gems/2.3.0/gems/chef-12.16.42/lib/chef/application.rb:59:in `run'
/opt/chef/embedded/lib/ruby/gems/2.3.0/gems/chef-12.16.42/bin/chef-client:26:in `<top (required)>'
/usr/bin/chef-client:54:in `load'
/usr/bin/chef-client:54:in `<main>'
```

```
[2016-11-25T17:15:25+03:00] DEBUG: chef_gem[pg] using gem '/opt/chef/embedded/bin/gem'
[2016-11-25T17:15:25+03:00] DEBUG: chef_gem[pg] found installed gem pg version 0.19.0 matching pg (>= 0)
[2016-11-25T17:15:25+03:00] DEBUG: chef_gem[pg] pg 0.19.0 already installed
[2016-11-25T17:15:25+03:00] DEBUG: chef_gem[pg] is already installed - nothing to do
[2016-11-25T17:15:25+03:00] DEBUG: chef_gem[pg] resetting gem environment to default
[2016-11-25T17:15:25+03:00] ERROR: Chef::Exceptions::ChildConvergeError: Chef run process terminated by signal 6 (IOT)
[2016-11-25T17:15:25+03:00] DEBUG: Chef::Exceptions::ChildConvergeError: Chef run process terminated by signal 6 (IOT)
```

```
[2016-11-25T17:30:28+03:00] INFO: Processing pgsql_user[test] action create (bosting-generator::default line 92)
[2016-11-25T17:30:28+03:00] DEBUG: Resources for generic postgresql_database_user resource enabled on node include: [Chef::Resource::PostgresqlDatabaseUser]
[2016-11-25T17:30:28+03:00] DEBUG: Resource for postgresql_database_user is Chef::Resource::PostgresqlDatabaseUser
[2016-11-25T17:30:28+03:00] INFO: Processing postgresql_database_user[test] action create (/etc/chef/local-mode-cache/cache/cookbooks/bosting-generator/resources/pgsql_user.rb line 11)
[2016-11-25T17:30:29+03:00] ERROR: Chef::Exceptions::ChildConvergeError: Chef run process terminated by signal 6 (IOT)
[2016-11-25T17:30:29+03:00] DEBUG: Chef::Exceptions::ChildConvergeError: Chef run process terminated by signal 6 (IOT)
```

**Comment by lamont-granquist** [comment]
This likely isn't a "bug" in this repo.  You can most likely reproduce the failure on the commandline as well if you put `/opt/chef/embedded/bin` in the PATH

The root cause of the bug is most likely solved by an approach similar to this pile of code, although I'm unfamiliar if this works on FreeBSD:

https://github.com/sous-chefs/postgresql-cookbook/blob/master/recipes/ruby.rb#L83-L143

Because the pg gem is such a phenomenally large pain in the ass to install as a chef_gem there's an accepted RFC to prebake it into the omnibus build:

https://github.com/chef/chef-rfc/blob/master/rfc063-omnibus-chef-native-gems.md

That has unfortunately been around for a year with no traction on it.  It's unlikely to get fixed in the short term since the branching mess in the chef/chef/omnibus/Gemfile creates too much friction.

Going to close this as ultimately a dup of RFC063.  If you still need short-term help you should ask on the the slack channel and/or stackoverflow.

**Comment by sekrett** [comment]
Thanks, I am happy to hear you are aware of it, but don't understand how rfc063 would help.

I fixed this problem but the steps were very strange. I deleted /root/.chef, /root/nodes, /etc/chef/local-mode-cache, did not help. Then I deleted /root/cookbooks, uploaded again and run chef-client non daemon version, it completed and after that again I can run daemon without problems. Maybe the reason of failure was the interruption on chef-client and some lock file was not cleaned up properly? That appears to be the only reasonable explanation to me, because I checked `ps ax` and also tried a reboot, it did not help.
